### ISSUE TITLE
idea/help: Better route(s) defining (and reusable components continuation)

### ISSUE THREAD

**Issue by tunnckoCore** [yoshuawuyts/choo#336]
Hi there. I'm just continue thinking about "reusable" components started from #266. And I got some structure and ideas.

So currently we can define only one time an array of routes, right? Using `app.router`. But why not add `app.route` method or something like that, to add routes more dynamically, it would be more flexible and will allow to define which component or page for which route is, inside from the component/page.

I'm thinking for the following thing

```
components/
- componentName/index.js
- componentName/model/
- componentName/model/index.js
- componentName/model/reducers.js
- componentName/model/effects.js
- componentName/model/state.js
pages/
- home.js
- about.js
elements/
- input.js
- button.js
```

Actually I don't have real app so I can't decide correctly, and there are some random thoughts in my head.

Description of above structure:
- `componentName/index.js` would be the view also it can be that mentioned `.choo` file. But really we don't need choo files, because they limiting us for no reason. We can have local state again. In that file we will just have `choo/html` require at least, or we can even require more **elements** (from elements folder) to construct some view or layout.
- `componentName/model/` - model things such as reducers and effects. It's intentional that it is a folder. That structure is inspired by @kaktus project, where parts of the model are separated. This also allows if some view/page/other component, want to use that model - like just `require('./model')` _(actually it won't be recommended / correct, because they would be loaded anyway and the way they should be used is `send('namespace:reducerOrEffect')`)_ 
- `componentName/model/index.js` just autogenerated file which is exactly like this [index](https://github.com/kaktus/kaktus/blob/b1b3351697540f81e65fc71c8e1a4da40f627541/src/app/models/general/index.js) - just exporting state, reducers and adds namespace. Optionally it can be just more flat if someone don't want to separate files for the model parts.

So what's related with the router. It's easy. Pages in another side would be more higher-level things assembling few components. So it would be useful if any page can define it's route, so should export function that gets `app` argument.

example about page
```js
module.export = function (app) {
  // assemble few components, layouts and etc
  app.route('/about', aboutPage)
}
```

The second thing which may should be changed is the "plugins", that i think it should accept function, that gets `app` argument and returns something. So you will have instance and would have ability to add things, routes, to register components on some cache (object or array) and etc.

So later you will have client.js like this

```js
const choo = require('choo')
const log = require('choo-log') // updated to new plugins api
// note: this not exist, but could be wrapper for choo
const app = require('choo-components')(choo({
  hooks: 'here'
}))

app.use(log())

// register components (may not make sense? not sure)
// app.use(require('components/form'))
// app.use(require('components/foobar'))
// app.use(require('components/google-maps'))

// or directly register `pages`
app.use(require('pages/home'))
app.use(require('pages/about'))

document.body.appendChild(app.start())
```

It's not very clear in my mind yet, but if we change plugins API we could do better things. And every page would just add it's route to which it should point.
Also one thing to mention is that maybe `app.start` should "run/start" the router (pass the `app.routes` array for example to `app.router`)

***

**So in short**, the things may look complex and complicated, but they aren't (everything other will be in separate wrapper lib - the things about components/pages loading and generating). The needed changes is the plugins api and add `app.route` to add single route to `app.routes` array that later is passed to `app.router` from the `app.start`

edit: It is easy to just create another method like `.component` instead of changing plugins api, but `use` is good word and make sense the plugins to be more better and still allow defining `hooks`.

**Comment by Ryuno-Ki** [comment]
How would you prevent route collisions?

**Comment by tunnckoCore** [comment]
Good and interesting question, but don't think it is "framework" job to do this. The mentioned `app.route` would be something like `app.routes.push(route)` thing, it's up to the user.

Also I don't think it is protected from this as it is currently. If user/dev want to break his job it can. The core thing behind "frameworks" should be just thin sugar and helper things with thin borders. We don't need strict rules.

**Comment by Ryuno-Ki** [comment]
I was thinking about namespacing the routes. That is, the call to `app.route` could maybe introspect from where it was called and incorporate that in the route.

For example, take the standard CRUD routes:
* `app.route('/create', CreateView)`
* `app.route('/:id', DetailView)`
* `app.route('/', ListView)`
* `app.route('/update/:id', UpdateView)`
* `app.route('/delete/:id', DeleteView)`

Would be cool, if one could write them like that.

**Comment by tunnckoCore** [comment]
That's good idea, yea, but

> could maybe introspect from where it was called and incorporate that in the route.

adds complexity and more detection things, it would become harder. But yea, one new method `app.route` gives such good looking route defining.

**Comment by Ryuno-Ki** [comment]
Sure, the trade-off is between sugar or configuration.

I would consider to have to think about all possible routes harder than a more complicate implementation. Keep in mind that you will face collisions as soon as you start integrating this routing component, because you cannot control the naming practices of other developers of the community.

That being said, how does other router implementations (Vue.js, React, Angular, Aurelia, …) solved this problem? Could you look for prior art?

**Comment by tunnckoCore** [comment]
> Keep in mind that you will face collisions as soon as you start integrating this routing component

Not sure, but don't see the problem, when you will define them from each component with full path. I think it would have more collisions if The Framework handles that things, rather then the user.

If you have nested components, and few components combines one component that is a Page (component) the only place where route is defined would be in that Page component/view.

I just don't like that "trie-based" routers maybe, don't know. But let's say we have one simple `app.routes` array. Which should contain objects with `path` and `component` (or `template`, `view` or something like that), then in the `app.start` will just loop them and fire the handlers. So then I don't see any possibility for collisions.

The Vue router is just some strange thing to me, currently. I was just looking over vue these last days. So what i seen last is `app.routes = [{ path: '/foo', component: Bar }, /* more routes */ ]`. But anyway, didn't look the implementation deeply.

For inspiration for good, ultra-composable and flexible route can look at [tunnckoCore/koa-better-router](https://github.com/tunnckoCore/koa-better-router) which is just absolutely thin layer on top of `path-to-regexp` - just more declarative syntax and more composable methods (most of the lines are comments) which in turn allowed me to built great koa-rest-router on top of it with few lines of code.

Btw, which reminds me that i should extract it from that koa, and create separate lib that would be just client and server router which will be able to configure what should be the arguments - in koa the route's handler will get `ctx, next` (comes from koa-compose in that case), and in other lib (for example choo, just example) will get `state, prev, send` args.

Anyway, i'm not promoting. Just saying that if we are not based on trie-based router we won't have collisions and confusion when using nested components.

~~Let's say we have nested route `/companies/:company/offices/:office/departments/:department` and we have separate "components" for each of them - company, office, department.~~

Actually, can you give some nested routes and examples? Cuz I always ends up with REST and I think it is another story and the things a bit more tricky.

**Comment by yoshuawuyts** [comment]
I started off with the idea of allowing each page to define their own routes but quickly realized that it allows hiding abstractions which makes for a good amount of complexity.

I feel what you're proposing is different enough that it would require a full rewrite - something that I can encourage you to do, and given choo's modular nature it should be very much possible to do.

Gonna close this for now tho, happy to hear how your experiments go!
